### PR TITLE
KAFKA-14680: Upgrade gradle version from 7.6 to 8.0.1

### PR DESCRIPTION
**Pull request by dejan2609**
Details:
 * gradle upgrade: 7.6 -> 8.0.1
 * spotbugs plugin upgrade: 5.0.9 -> 5.0.13
 * tweaked the mechanics for `-release`/`-source`/`-target` to workaround idiosyncrasies in Gradle 8.0.1 and newer Scala 2.13 versions.
 * streams-scala `test` task no longer triggers the `spotless` task since a newer version is required for Gradle 8 support, but the newer version requires Java 11.
Note: relates to #5479

Gradle upgrade highlights:
* "Scala Incremental Compilation for Multi-Module projects broken in 7.x": https://github.com/gradle/gradle/issues/20101
* "Incremental compilation of java modules is broken with Gradle 7.6": https://github.com/gradle/gradle/issues/23067

Full release notes: https://docs.gradle.org/8.0/release-notes.html

### CODE REVIEW

**Comment by dejan2609** [comment]
Related to #13199 

Thanx for heads up @ijuma 

On my local box build ends up in green, not sure why it fails on Jenkins; will investigate (and wait for plugin changes to be merged).


```
18:46:36  + ./retry_zinc ./gradlew -PscalaVersion=2.13 clean compileJava compileScala compileTestJava compileTestScala spotlessScalaCheck checkstyleMain checkstyleTest spotbugsMain rat --profile --continue -PxmlSpotBugsReport=true -PkeepAliveMode=session
18:46:36  Error: Could not find or load main class org.gradle.wrapper.GradleWrapperMain
18:46:36  Caused by: java.lang.ClassNotFoundException: org.gradle.wrapper.GradleWrapperMain
```

**Comment by ijuma** [comment]
@dejan2609 We'll wait for the final release, but it's fine to start working through the issues. Looks like the build failed.

**Comment by dejan2609** [comment]
@ijuma yes, somehow build  on Jenkins fails to start, I will rebase and then test on my local box (with Windows/Linux).

**Comment by ijuma** [comment]
Gradle 8 has been released. Have you been able to figure out why the build is failing?

**Comment by dejan2609** [comment]
@ijuma some findigs: build fails due to issues with gradle wrapper bootstrapping. 

I will post more details today.

**Comment by dejan2609** [comment]
Update: gradle wrapper bootstrapping is ok now, but spotless Scala checks are failing... 
Searching for a solution.

**Comment by dejan2609** [comment]
It seems that Spotless Gradle plugin needs to be alligned with Gradle 8.0 (I filed a ticket here: https://github.com/diffplug/spotless/issues/1572)

Thing is that they dropped support for direct Java 8 builds: https://github.com/diffplug/spotless/blob/main/plugin-gradle/CHANGES.md#6140---2023-01-26 (their suggestion for other teams is to use Java cross compilation: https://docs.gradle.org/8.0/userguide/building_java_projects.html#sec:java_cross_compilation).

Kafka obviously still needs to build artifacts against Java 8, so maybe it would be a good idea to use Spotless team suggestion.

All-in-all, herewith a plan for a Gradle 7 -->> 8 upgrade:
- spotless team will release Gradle 8.0 compatible version (most probably they will not backport solution into spotless gradle 6.13.x line)
- in parallel I can try to drop JDK 8 usage (note: Java 8 compatible artifacts will still be generated)
- when we make sure that Java cross compilation works we can come back to this PR and bump Gradle (and Spotless plugin version to 6.15+).

@ijuma If it is ok with you I can start working towards this solution.

**Comment by ijuma** [comment]
An alternative would be to drop spotless until we drop support for Java 8 (Apache Kafka 4.0). What actually uses spotless today?

**Comment by dejan2609** [comment]
At the moment spotless is used via Jenkins CI server :arrow_right: Jenkinsfile 'spotlessScalaCheck' task execution: https://github.com/apache/kafka/blob/3.4.0/Jenkinsfile#L23

My suggestion is to:
- remove 'spotlessScalaCheck' out of Jenkinsfile 
- put some comments into build.gradle that would prevent others from changing anything related to spotless until Kafka 4.0 comes into play (i.e. after Java 8 is dropped):
   - plugin definition: https://github.com/apache/kafka/blob/3.4.0/build.gradle#L33
   - plugin configuration: https://github.com/apache/kafka/blob/3.4.0/build.gradle#L47

Note: this PR already removes task [dependency](https://github.com/apache/kafka/blob/3.4.0/build.gradle#L2084).

**Comment by dejan2609** [comment]
🤔 Interesting, altough 'spotlessScalaCheck' is removed out of Jenkinsfile that task is still executed on Jenkins: 

image

**Comment by ijuma** [comment]
Yeah, that's a security feature. I suggest filing a JIRA about disabling spotlessScalaCheck with the rationale and have a separate PR for that change.

**Comment by dejan2609** [comment]
Update: there are some new issues with Gradle 8.0.1; I will try to investigate that also.

**Comment by dejan2609** [comment]
@ijuma Good news, it seems that I managed to provide a workaround for Gradle 8.0.1 (and across all JDK/Scala versions that Apache Kafka uses and the moment).
I still need to test few more time and after that will move this PR from _draft_ to _ready for review_.

**Big** shoutout  for @ljacomet from Gradle team, he went over and above during investigation 🎉

**Comment by ijuma** [comment]
Thanks @dejan2609. I still prefer the path I described here:

> https://github.com/gradle/gradle/issues/23962#issuecomment-1439115365

Do you know if that works?

**Comment by inglor** [comment]
> Thanks @dejan2609. I still prefer the path I described here:
> 
> > [gradle/gradle#23962 (comment)](https://github.com/gradle/gradle/issues/23962#issuecomment-1439115365)
> 
> Do you know if that works?

Hi @ijuma , 

just to pitch in that setting `-release` on scala `2.13` is producing same error
`[Error] : -target is deprecated: Use -release instead to compile against the correct platform API.`

```
diff --git a/build.gradle b/build.gradle
--- a/build.gradle	(revision 2e1947d240607d53f071f61c875cfffc3fec47fe)
+++ b/build.gradle	(date 1677025069098)
@@ -684,7 +684,7 @@
     }
 
     // Scalac's `-release` requires Java 9 or higher
-    if (JavaVersion.current().isJava9Compatible())
+    if (versions.baseScala == '2.13')
       scalaCompileOptions.additionalParameters += ["-release", minJavaVersion]
 
     configure(scalaCompileOptions.forkOptions) {

```

PS: Encountered this error while trying to package `kafka` for Arch Linux which has latest gradle version `8.0.1`

**Comment by ijuma** [comment]
@inglor Thanks. The following comment explains that the way `-release` must be set is a bit specific:

https://github.com/gradle/gradle/issues/23962#issuecomment-1437348400

**Comment by dejan2609** [comment]
@inglor thanx for stopping by, any insight or help is highly appreciated ! 

@ijuma I scraped fast solution that, suprisingly enough, works (at least compilation is succesful, that is). 

I tested locally across entire matrix: JDK 8/11/17 x Scala 2.12/13 (with ./gradlew -PscalaVersion=2.1X clean jar) and all builds are fine. 

Still need to spend some time with this to doublecheck and maybe even to compare compilation console output (trunk vs. this PR).

**Comment by inglor** [comment]
@ijuma Thanks for the pointer - I have missed that.
@dejan2609 I can confirm your cdbb59fad79b4296b2095417274050b7f1ea1698 works fine (as in it compiles the scala code) for

**Gradle version 8.0.1:**

|            | Scala 2.12       | Scala 2.13       |
|------------|------------------|------------------|
| JDK 8      |:white_check_mark:|:white_check_mark:|
| JDK 11     |:white_check_mark:|:white_check_mark:|
| JDK 17     |:white_check_mark:|:white_check_mark:|

**Comment by ijuma** [comment]
@dejan2609 I added a few cleanups here: https://github.com/ijuma/kafka/commit/9c6ae575276e673f39d188d42a332f9f0b07d2d0

If you agree with them, please integrate into your PR.

**Comment by dejan2609** [comment]
Got it @ijuma 

I will also add  jacoco minor version bump: 0.8.7 -->> 0.8.8 in order to use same version as in Gradle 8.0.1:
https://docs.gradle.org/8.0.1/userguide/jacoco_plugin.html#sec:configuring_the_jacoco_plugin

Will squash/rebase on trunk and force-push.

**Comment by dejan2609** [comment]
> @dejan2609 I added a few cleanups here: [ijuma@9c6ae57](https://github.com/ijuma/kafka/commit/9c6ae575276e673f39d188d42a332f9f0b07d2d0)
> 
> If you agree with them, please integrate into your PR.

@ijuma Scala 2.13 builds are fine, but something is still missing for Scala 2.12 (all **_./gradlew -PscalaVersion=2.12 clean jar_** builds are failling):
- JDK 17
```
> Task :core:compileScala
'-release' does not accept multiple arguments
bad option: '-release:8'

> Task :core:compileScala FAILED

> Task :streams:streams-scala:compileScala FAILED
'strict-unsealed-patmat' is not a valid choice for '-Xlint'
'-release' does not accept multiple arguments
bad option: '-release:8'

FAILURE: Build completed with 2 failures.
```
- JDK 11
```
> Task :core:compileScala FAILED

> Task :streams:streams-scala:compileScala
'strict-unsealed-patmat' is not a valid choice for '-Xlint'
'-release' does not accept multiple arguments
bad option: '-release:8'

> Task :streams:streams-scala:compileScala FAILED

FAILURE: Build completed with 2 failures.
```
- JDK 8
```
> Task :core:compileScala FAILED
'-release' does not accept multiple arguments
bad option: '-release:8'

FAILURE: Build failed with an exception.
```

**Comment by ijuma** [comment]
@dejan2609 Good catch. I fixed it in the same branch. And tested with Scala 2.12 and 2.13. See https://github.com/ijuma/kafka/commit/31aec82fe7077c790f79c94f91a4908b03ae6556

**Comment by dejan2609** [comment]
@ijuma changes are integrated (please remember that https://github.com/apache/kafka/pull/13263 needs to be merged prior to this one).

**Comment by ijuma** [comment]
@dejan2609 I merged #13263 and updated the PR description. Please take a look.

**Comment by ijuma** [comment]
@dejan2609 One issue is that IntelliJ still doesn't handle this correctly. Before this PR, it would add the following to `.idea/scala_compiler.xml`:

> ``<parameter value="-target:jvm-1.8" />``

Now it adds:

> ``<parameter value="-target:jvm-17" />``

Both cause the build to break with the latest Scala 2.13 version since that deprecated `-target` and we compile with fatal warnings.

File: gradlew
Comment (ijuma) [inline gradlew]: Something weird happened here. When I run `./gradlew wrapper on my computer, it generates the following diff:

image

Comment (ijuma) [reply gradlew]: If you can fix this, I think we can merge this PR.

**Comment by ijuma** [comment]
I pushed a commit that reverts part of my previous commit. We need to set `sourceCompatibility` and `targetCompatibility` to workaround an issue affecting the Scala plugin for IntelliJ. We're back to with this change:

> ``<parameter value="-target:jvm-1.8" />``

This is still a problem since this is deprecated in newer Scala versions, but this PR doesn't make it worse.

@dejan2609 would you be willing to submit a ticket for the Scala plugin for IntelliJ so it doesn't set `-target` if `-release` is set?

File: gradlew
Comment (ijuma) [reply gradlew]: I pushed an additional commit to fix this.

**Review by ijuma — APPROVED** [review]
LGTM, thanks. I'll wait for the CI to complete before merging. The last CI build looked good, so it should be fine.

File: gradlew
Comment (dejan2609) [reply gradlew]: From the top of my head this may be related with zero width characters or something od that sort (https://github.com/microsoft/terminal/issues/11850) 

Thing is that I am constantly switching from Windows to Linux and back, and this commit was created in Windows.

In any case, thank you for solving this.

**Comment by dejan2609** [comment]
> @dejan2609 would you be willing to submit a ticket for the Scala plugin for IntelliJ so it doesn't set `-target` if `-release` is set?
@ijuma  Sure, will do that.

**Comment by ijuma** [comment]
JDK 8 build passed, the other two had unrelated failures:

> Build / JDK 17 and Scala 2.13 / org.apache.kafka.streams.processor.internals.DefaultStateUpdaterTest.shouldPauseActiveTaskAndTransitToUpdateStandby()
Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.integration.ConnectorRestartApiIntegrationTest.testMultiWorkerRestartOnlyConnector
Build / JDK 11 and Scala 2.13 / org.apache.kafka.tools.MetadataQuorumCommandTest.[5] Type=Raft-CoReside, Name=testDescribeQuorumStatusSuccessful, MetadataVersion=3.5-IV0, Security=PLAINTEXT

**Comment by chia7712** [comment]
Should we disable `spotlessScala` temporarily? otherwise, the Gradle task related to `spotlessScala` can't work. For example, the common command `./gradlew clean build -x test`.

**Comment by dejan2609** [comment]
@chia7712 Correct, `./gradlew build` executed against trunk does fail.

For the last few days we (@ijuma and me):
- merged into a trunk Jenkinsfile change that removes 'spotlessScalaCheck': https://github.com/apache/kafka/commit/97efdc65f082d72b7980e6a8b1d56ba0d70002fd 
- merged this PR also (that was tested with `./gradlew jar` on our local machines and then on Jenkins)

However,  `./gradlew build` depends on spotless tasks (tree is generated using this Gradle plugin: https://plugins.gradle.org/plugin/org.barfuin.gradle.taskinfo)

`./gradlew build tiTree` 

```
> Task :tiTree
:build                                                         (org.gradle.api.DefaultTask)
+--- :assemble                                                 (org.gradle.api.DefaultTask)
|    `--- :jar                                                 (org.gradle.api.tasks.bundling.Jar)
|         +--- :classes                                        (org.gradle.api.DefaultTask)
|         |    +--- :compileJava                               (org.gradle.api.tasks.compile.JavaCompile)
|         |    `--- :processResources                          (org.gradle.language.jvm.tasks.ProcessResources)
|         `--- :compileJava                                    (org.gradle.api.tasks.compile.JavaCompile)
`--- :check                                                    (org.gradle.api.DefaultTask)
     +--- :rat                                                 (org.nosphere.apache.rat.RatTask)
     +--- :spotlessCheck                                       (org.gradle.api.DefaultTask)
     |    `--- :spotlessScalaCheck                             (com.diffplug.gradle.spotless.SpotlessCheck)
     |         `--- :spotlessScala                             (com.diffplug.gradle.spotless.SpotlessTaskImpl)
     |              `--- :spotlessInternalRegisterDependencies (com.diffplug.gradle.spotless.RegisterDependenciesTask)
     `--- :test                                                (org.gradle.api.tasks.testing.Test)
          +--- :classes                                        (org.gradle.api.DefaultTask)
          |    +--- :compileJava                               (org.gradle.api.tasks.compile.JavaCompile)
          |    `--- :processResources                          (org.gradle.language.jvm.tasks.ProcessResources)
          +--- :compileJava                                    (org.gradle.api.tasks.compile.JavaCompile)
          +--- :compileTestJava                                (org.gradle.api.tasks.compile.JavaCompile)
          |    +--- :classes                                   (org.gradle.api.DefaultTask)
          |    |    +--- :compileJava                          (org.gradle.api.tasks.compile.JavaCompile)
          |    |    `--- :processResources                     (org.gradle.language.jvm.tasks.ProcessResources)
          |    `--- :compileJava                               (org.gradle.api.tasks.compile.JavaCompile)
          `--- :testClasses                                    (org.gradle.api.DefaultTask)
               +--- :compileTestJava                           (org.gradle.api.tasks.compile.JavaCompile)
               |    +--- :classes                              (org.gradle.api.DefaultTask)
               |    |    +--- :compileJava                     (org.gradle.api.tasks.compile.JavaCompile)
               |    |    `--- :processResources                (org.gradle.language.jvm.tasks.ProcessResources)
               |    `--- :compileJava                          (org.gradle.api.tasks.compile.JavaCompile)
               `--- :processTestResources                      (org.gradle.language.jvm.tasks.ProcessResources)
```

**Comment by chia7712** [comment]
@dejan2609 thanks for nice explanation!

That is why I suggested to disable spotlessScala from build.gradle :)

**Comment by dejan2609** [comment]
@chia7712 I just pushed PR that (hopefully) solves this regression, feel free to check it.

**Comment by dejan2609** [comment]
> @dejan2609 would you be willing to submit a ticket for the Scala plugin for IntelliJ so it doesn't set `-target` if `-release` is set?

@ijuma ticket is created here: https://youtrack.jetbrains.com/issue/SCL-21039

**Comment by dejan2609** [comment]
@ijuma please review Scala plugin for Intellij ticket (link above); I need some help to provide reproducer.

**Comment by ijuma** [comment]
The `spotlessScala` issue was fixed via #13311.

**Pull request merged**
